### DISCUSSION TITLE
How to use just the recognition model with a custom object detection crop?

### DISCUSSION THREAD

**Discussion by ajkdrag** [mindee/doctr#1463] [Q&A] [answered]
I am doing object detection with a different framework, which gives me the bboxes for few text regions i am interested in. When I run just the recognition_predictor of DocTR, the output is gibberish. Am I missing something here? When I run ocr_predictor on the entire original image, the detection and recognition results seem to look fine.

**Comment by felixdittrich92** [comment]
Hi @ajkdrag :wave:

Could you provide a short snippet ? :)

**Reply by ajkdrag** [reply]
Hi, looks like i was trying to use the Parseq model, but since the pretrained URL isn't available yet, so it was giving me gibberish.

If I ignore Parseq, which is the current best recog model for texts, provided by DocTR?
If I want to use Parseq with release 0.7, how can I do so? Do I have to use the main branch and install from there?
Thanks!

**Reply by felixdittrich92** [reply, marked as answer]
1. with 0.7.0 it is crnn_vgg16_bn or crnn_mobilenet_v3_large
2. correct you would need to install from the main branch (we will do the 0.8.0 release soon) :)

**Reply by ajkdrag** [reply]
Thanks a lot! In the demo, so far I observed parseq doing better on documents. Not sure if that's expected?

**Reply by felixdittrich92** [reply]
All doctr models are trained from scratch on a mindee internal dataset which contains documents, receipts, etc. :) 
So the models are especially trained to work well on textual rich images / documents 

But yeah parseq is also from benchmarks our best performing model atm (inference latency and accuracy) 
Only master reaches small better scores but it's also super slow and large ^^

**Reply by ajkdrag** [reply]
Thanks for the response. Understood!

**Reply by ajkdrag** [reply]
If I use the main branch, is there any breaking change?

**Reply by felixdittrich92** [reply]
A minor one we changed

`result.show(doc)`

to 

`result.show()`

otherwise no breaking changes only improvements ^^

**Reply by ajkdrag** [reply]
One final question here:
I ran just the detection model (db_resnet50) from latest release (not the main branch), and compared the bboxes visually against the huggingface spaces drawn bboxes. For few images, there are subtle but important differences. The bboxes in the hf spaces demo seem to chunk words better. Is this normal? Should I use the main branch checkpoints for better boxes?

**Reply by felixdittrich92** [reply]
Hi @ajkdrag correct :+1: 
We have retrained all models (only TF db_resnet50 / db_mobilenet_v3_large are left) - only on the main branch available yet :)

**Reply by ajkdrag** [reply]
Any idea when the pytorch will be available? For my usecase, db_resnet50 performs well (the retrained one), but the latency is very high. Any tips to reduce this? I am aiming to run on cpu with just the detection_predictor interface. I will use a different recognition framework.

**Reply by felixdittrich92** [reply]
I hope that the last both runs goes well and we can release between the next 7-14 days

Mh on CPU there is not much space (you could play a bit with the `det_bs` depending on your hardware and batches you pass trough the model)

With a nvidia gpu and half precision you could reduce the runtime by up to 80%

**Reply by ajkdrag** [reply]
Thanks. I was trying to set the det_bs, but couldn't find the param for it the documentation.
I am using the detection_predictor() interface, which takes in the model and the inputs. How do i pass the det_bs arg?

**Reply by felixdittrich92** [reply]
```
from doctr.models.detection.predictor import DetectionPredictor
from doctr.models.preprocessor import PreProcessor
from doctr.io import DocumentFile

doc = DocumentFile.from_pdf("path/to/your/doc.pdf")

det_predictor = DetectionPredictor(
    PreProcessor(
        (1024, 1024),
        batch_size=1,  # set whatever your hardware can handle
        mean=(0.798, 0.785, 0.772),
        std=(0.264, 0.2749, 0.287)
    ),
    "db_resnet50"
)

preds = det_predictor(doc)
```

NOTE: box sorting is now on your own without using the OCRPredictor` instance :)

**Reply by felixdittrich92** [reply]
Can i ask which recognition model / framework you will use and why ? language not supported ? :)

**Reply by ajkdrag** [reply]
I tried Google's OCR which does really well, but isn't free/open source. 
I also tried tesseract which is surprisingly doing well too ( i used db_resnet50 boxes with tesseract best data and compared it with crnn_vgg16_bn) and tesseract was marginally better (if i used --psm 7 or --psm 8). For some reason the db_resnet50 boxes still have two words together clubbed *sometimes* (even if i use main branch). 
I will also try paddleOCR's recognition with doctr boxes to see if it does well.
As of now, after testing with docTR's mobilenet large recognition models, the recognition seems well, but still Google's OCR is better. Idk what model they use. 
Do you suggest I train the recognition model on my data? Language is english and the documents are official cheque docs. Is it worth it? I only have like 150 images.

**Reply by felixdittrich92** [reply]
Yeah you could fine tune the model on your data 80 / 20 split :)
Would it be possible to provide an example image to see about "clubbed sometimes" ?

**Reply by ajkdrag** [reply]
"County Treasurer" is merged into one bbox. Now maybe here the spacing is too tight?

![1707921396647](https://github.com/mindee/doctr/assets/40544701/994a4a29-182f-438b-8514-2197e4139d7d)
Also, is there a way to finetune the detector model to only give boxes in ROI i am interested in? I saw there's a maskrcnn and classification modules in DocTR but didn't find further info?

**Reply by felixdittrich92** [reply]
The Mask-RCNN is used for object detection (qrcodes, passphoto, etc.).
But it is deprecated in my mind and not really maintained so i think we will drop it in the next releases and implement something comparable but better working.

The classification models are used as our backbones for the detection and recognition models mainly. (and for orientation classification)

Have you tried to increase the binarization threshold to get more accure predictions ? 

For example:
```
det_predictor.model.postprocessor.bin_thresh = 0.5
det_predictor.model.postprocessor.box_thresh = 0.2
```

**Reply by ajkdrag** [reply]
The bin_thresh does manage to split, but many boxes are lost in the process. I can evaluate and check if this helps. Thanks!

**Reply by ajkdrag** [reply]
How does the KIE predictor work? If only want to capture detections/ocr from certain regions of the image,will it help this usecase?

**Reply by felixdittrich92** [reply]
Yeah but in this case you would need to fine tune a model.
Before you start with this i would try to manipulate the detection predictions in the middle of the pipeline with our hook: (only main branch atm)

https://mindee.github.io/doctr/latest/using_doctr/using_models.html#advanced-options

This will still process the whole image but you could for example exclude all detection predictions (boxes) which are not in your known ROI

Hope this helps :)

**Reply by ajkdrag** [reply]
Nice, i was trying the hooks, but i was using the 0.7 release branch. Thanks, this helps!

**Reply by ajkdrag** [reply]
I was using the document builder to merge words into lines:
![image](https://github.com/mindee/doctr/assets/40544701/56bba8c2-dea6-4551-bd65-c505fc292615)
As you can see above, the lines "pay against this cheque" get merged with "Mejeb khalid alhajri". Is there a parameter to control this behaviour? Ideally it should be separate.

**Reply by ajkdrag** [reply]
i used  paragraph_break=0.02 to fix this. Thanks. DocTR is pretty cool.

**Reply by ajkdrag** [reply]
I will post my queries in separate discussion, so that all these insights doesn't get lost in a single thread, and others can find help too.

**Reply by felixdittrich92** [reply]
Nice that it works now 😊
Oh yeah i see we should anyway extend the docs especially about the DocumentBuilder options cc @odulcy-mindee

**Reply by vikasr111** [reply]
@ajkdrag I am curious how you managed to merge the words into the lines. I am facing the same problem where the DocTR output shows words from same lines into different lines. I was looking for a reliable way to do line parsing so that the lines in the output are similar as present in the document.

I tried several methods but none of them worked. I noticed in your image that you have managed to find a solution.

**Comment by ajkdrag** [comment]
> @ajkdrag I am curious how you managed to merge the words into the lines. I am facing the same problem where the DocTR output shows words from same lines into different lines. I was looking for a reliable way to do line parsing so that the lines in the output are similar as present in the document.
> 
> I tried several methods but none of them worked. I noticed in your image that you have managed to find a solution.

`paragraph_break=0.02` is the param you can try to adjust that.
